### PR TITLE
Bug 1719724 Add labeling of click spikes for contextual services

### PR DESCRIPTION
**Pull request by jklukas**
See https://bugzilla.mozilla.org/show_bug.cgi?id=1719724

### CODE REVIEW

File: ingestion-beam/src/main/java/com/mozilla/telemetry/contextualservices/LabelClickSpikes.java [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+/**
+ * Transform that maintains state per key in order to label suspicious clicks.
+ */
+public class LabelClickSpikes extends
+    PTransform<PCollection<KV<String, PubsubMessage>>, PCollection<KV<String, PubsubMessage>>> {
+
+  private final Integer maxClicks;
+  private final Long windowMillis;
+  private final Counter status64Counter = Metrics.counter(LabelClickSpikes.class,
+      "click_status_64");
```
Comment (jklukas) [inline ingestion-beam/src/main/java/com/mozilla/telemetry/contextualservices/LabelClickSpikes.java:45]:
```suggestion
  private final Counter status65Counter = Metrics.counter(LabelClickSpikes.class,
      "click_status_65");
```

File: ingestion-beam/src/main/java/com/mozilla/telemetry/contextualservices/LabelClickSpikes.java [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    state.write(timestamps);
+    return timestamps;
+  }
+
+  /** Updates the passed attribute map, adding click-status to the reporting URL. */
+  private static void addClickStatusToReportingUrlAttribute(Map<String, String> attributes) {
+    String reportingUrl = attributes.get(Attribute.REPORTING_URL);
+    ParsedReportingUrl urlParser = new ParsedReportingUrl(reportingUrl);
+    // "64" indicates a spike in clicks per the Conducive API spec.
+    urlParser.addQueryParam("click-status", "64");
```
Comment (jklukas) [inline ingestion-beam/src/main/java/com/mozilla/telemetry/contextualservices/LabelClickSpikes.java:91]:
```suggestion
    // "65" indicates a short-term spike in "ghost" clicks per the Conducive API spec.
    urlParser.addQueryParam("click-status", "65");
```

File: ingestion-beam/src/main/java/com/mozilla/telemetry/contextualservices/LabelClickSpikes.java [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+      // it will overwrite this timer value.
+      timer.set(Instant.ofEpochMilli(elementTs.getMillis() + windowMillis));
+
+      if (timestamps.size() <= maxClicks) {
+        out.output(element);
+      } else {
+        PubsubMessage message = element.getValue();
+        Map<String, String> attributes = new HashMap<>(message.getAttributeMap());
+        addClickStatusToReportingUrlAttribute(attributes);
+        status64Counter.inc();
```
Comment (jklukas) [inline ingestion-beam/src/main/java/com/mozilla/telemetry/contextualservices/LabelClickSpikes.java:120]:
```suggestion
        status65Counter.inc();
```

File: ingestion-beam/src/test/java/com/mozilla/telemetry/contextualservices/LabelClickSpikesTest.java [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+    PAssert.that(result).satisfies(iter -> {
+      int size = Iterables.size(iter);
+      assert size == 40 : "Expected 40 messages, but found " + size;
+      return null;
+    });
+
+    PAssert.that(result).satisfies(iter -> {
+      long countWithStatus = StreamSupport.stream(iter.spliterator(), false) //
+          .filter(m -> m.getAttribute(Attribute.REPORTING_URL).contains("click-status=64")) //
```
Comment (jklukas) [inline ingestion-beam/src/test/java/com/mozilla/telemetry/contextualservices/LabelClickSpikesTest.java:67]:
```suggestion
          .filter(m -> m.getAttribute(Attribute.REPORTING_URL).contains("click-status=65")) //
```

File: ingestion-beam/src/test/java/com/mozilla/telemetry/contextualservices/LabelClickSpikesTest.java [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+
+    PAssert.that(result).satisfies(iter -> {
+      int size = Iterables.size(iter);
+      assert size == 16 : "Expected 16 messages, but found " + size;
+      return null;
+    });
+
+    PAssert.that(result).satisfies(iter -> {
+      long countWithStatus = StreamSupport.stream(iter.spliterator(), false) //
+          .filter(m -> m.getAttribute(Attribute.REPORTING_URL).contains("click-status=64")) //
```
Comment (jklukas) [inline ingestion-beam/src/test/java/com/mozilla/telemetry/contextualservices/LabelClickSpikesTest.java:103]:
```suggestion
          .filter(m -> m.getAttribute(Attribute.REPORTING_URL).contains("click-status=65")) //
```

File: ingestion-beam/src/test/java/com/mozilla/telemetry/contextualservices/LabelClickSpikesTest.java [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+  @Rule
+  public TestPipeline pipeline = TestPipeline.create();
+
+  @Test
+  public void testSetsClickStatus() {
+    ImmutableMap<String, String> attributes = ImmutableMap.of(Attribute.CONTEXT_ID, "a", //
+        Attribute.REPORTING_URL, "https://test.com");
+    PubsubMessage message = new PubsubMessage(new byte[] {}, attributes);
+    PubsubMessage[] messages = new PubsubMessage[19];
+    Arrays.fill(messages, message);
```
Comment (BenWu) [inline ingestion-beam/src/test/java/com/mozilla/telemetry/contextualservices/LabelClickSpikesTest.java:33]: This array isn't used anywhere

File: ingestion-beam/src/test/java/com/mozilla/telemetry/contextualservices/LabelClickSpikesTest.java [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+    }
+
+    // Now, we add 20 messages each only a second apart. The first 10 should saturate the timestamp
+    // state, then the final 10 should be marked as suspicious via click-status.
+    for (int i = 1; i <= 20; i++) {
+      eventBuilder = eventBuilder
+          .advanceWatermarkTo(
+              new Instant(0L).plus(Duration.standardMinutes(30)).plus(Duration.standardSeconds(i)))
+          .addElements(message);
+    }
```
Comment (BenWu) [inline ingestion-beam/src/test/java/com/mozilla/telemetry/contextualservices/LabelClickSpikesTest.java:51]: I would like to see these as two separate test cases to be explicit about what's be marked with click status

**Review by BenWu — COMMENTED** [review]
I had to read up on beam states but this looks good to me.  I just have some minor comments and questions.

File: ingestion-beam/src/main/java/com/mozilla/telemetry/contextualservices/LabelClickSpikes.java [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+    attributes.put(Attribute.REPORTING_URL, urlParser.toString());
+  }
+
+  private class Fn extends DoFn<KV<String, PubsubMessage>, KV<String, PubsubMessage>> {
+
+    // See https://beam.apache.org/documentation/programming-guide/#state-and-timers
+    @StateId("click-state")
+    private final StateSpec<ValueState<List<Long>>> clickState = StateSpecs.value();
+    @TimerId("click-timer")
+    private final TimerSpec clickTimer = TimerSpecs.timer(TimeDomain.EVENT_TIME);
```
[the remark this answers is not in the document]
Comment (jklukas) [reply ingestion-beam/src/main/java/com/mozilla/telemetry/contextualservices/LabelClickSpikes.java:100]: I think you're right here. We're in kind of a squishy space as to what event time means for these clicks; the two concepts are not very different in our case, but processing time is indeed easier to reason about in general. Will change.

File: ingestion-beam/src/main/java/com/mozilla/telemetry/contextualservices/LabelClickSpikes.java [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+      timer.set(Instant.ofEpochMilli(elementTs.getMillis() + windowMillis));
+
+      if (timestamps.size() <= maxClicks) {
+        out.output(element);
+      } else {
+        PubsubMessage message = element.getValue();
+        Map<String, String> attributes = new HashMap<>(message.getAttributeMap());
+        addClickStatusToReportingUrlAttribute(attributes);
+        ghostClickCounter.inc();
+        out.output(KV.of(element.getKey(), new PubsubMessage(message.getPayload(), attributes)));
```
[the remark this answers is not in the document]
Comment (jklukas) [reply ingestion-beam/src/main/java/com/mozilla/telemetry/contextualservices/LabelClickSpikes.java:120]: Yes, we could. It feels more composable to have the input and output types be identical here, and have the normal case be that the element is passed through unaltered, then strip off the key in a separate step. In reality, it's unlikely we will have any need to use this in a context where we don't strip out the key immediately after the transform. I'll think a bit more about this.

[the remark this answers is not in the document]
Comment (jklukas) [reply ingestion-beam/src/main/java/com/mozilla/telemetry/contextualservices/LabelClickSpikes.java:120]: I think my strongest argument here is that the `LabelClickSpikes.perContextId` factory method ends up having a parallel structure as well. It takes in an unkeyed PCollection, handles wrapping it in a KV, explicitly handles unwrapping it (via `Values.create()`), and returns an unkeyed PCollection.

[the remark this answers is not in the document]
Comment (BenWu) [reply ingestion-beam/src/main/java/com/mozilla/telemetry/contextualservices/LabelClickSpikes.java:120]: Ok sounds reasonable, bit of a nit anyway

**Review by BenWu — APPROVED** [review]
Looks good.  Just wondering about the `advanceWatermarkTo` in the test

File: ingestion-beam/src/test/java/com/mozilla/telemetry/contextualservices/LabelClickSpikesTest.java [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+    ImmutableMap<String, String> attributes = ImmutableMap.of(Attribute.CONTEXT_ID, "a", //
+        Attribute.REPORTING_URL, "https://test.com");
+    PubsubMessage message = new PubsubMessage(new byte[] {}, attributes);
+    Builder<PubsubMessage> eventBuilder = TestStream.create(PubsubMessageWithAttributesCoder.of());
+
+    // We add 20 messages each only a second apart. The first 10 should saturate the timestamp
+    // state, then the final 10 should be marked as suspicious via click-status.
+    for (int i = 1; i <= 20; i++) {
+      eventBuilder = eventBuilder
+          .advanceWatermarkTo(new Instant(0L).plus(Duration.standardSeconds(i)))
```
[the remark this answers is not in the document]
Comment (jklukas) [reply ingestion-beam/src/test/java/com/mozilla/telemetry/contextualservices/LabelClickSpikesTest.java:38]: Good call. I'll update these to `advanceProcessingTime`. It's not clear to me why `advanceWatermarkTo` still works here, but may be an artifact of how `TestStream` is implemented.

**Pull request merged**